runtime-modules/storage/src/lib.rs

```diff
@@ -1,20 +1,2373 @@
+                bucket.voucher = Voucher{
+                    size_limit: new_objects_size_limit,
+                    objects_limit: new_objects_number_limit,
+                    ..bucket.voucher
+                };
+            });
+
+            Self::deposit_event(
+                RawEvent::StorageBucketVoucherLimitsSet(
+                    storage_bucket_id,
+                    worker_id,
+                    new_objects_size_limit,
+                    new_objects_number_limit
+                )
+            );
+        }
+
+        /// A storage provider signals that the data object was successfully uploaded to its storage.
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn accept_pending_data_objects(
+            origin,
+            worker_id: WorkerId<T>,
+            storage_bucket_id: T::StorageBucketId,
+            bag_id: BagId<T>,
+            data_objects: BTreeSet<T::DataObjectId>,
+        ) {
+            T::ensure_worker_origin(origin, worker_id)?;
+
+            let bucket = Self::ensure_storage_bucket_exists(&storage_bucket_id)?;
+
+            Self::ensure_bucket_invitation_accepted(&bucket, worker_id)?;
+
+            Self::validate_accept_pending_data_objects_params(
+                &bag_id,
+                &data_objects,
+                &storage_bucket_id
+            )?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            for data_object_id in data_objects.iter() {
+                BagManager::<T>::accept_data_objects(&bag_id, &data_object_id);
+            }
+
+            Self::deposit_event(
+                    RawEvent::PendingDataObjectsAccepted(storage_bucket_id, worker_id, bag_id, data_objects)
+            );
+        }
+
+        /// Update whether new bags are being accepted for storage.
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn update_storage_bucket_status(
+            origin,
+            worker_id: WorkerId<T>,
+            storage_bucket_id: T::StorageBucketId,
+            accepting_new_bags: bool
+        ) {
+            T::ensure_worker_origin(origin, worker_id)?;
+
+            let bucket = Self::ensure_storage_bucket_exists(&storage_bucket_id)?;
+
+            Self::ensure_bucket_invitation_accepted(&bucket, worker_id)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            <StorageBucketById<T>>::mutate(storage_bucket_id, |bucket| {
+                bucket.accepting_new_bags = accepting_new_bags;
+            });
+
+            Self::deposit_event(
+                RawEvent::StorageBucketStatusUpdated(
+                    storage_bucket_id,
+                    worker_id,
+                    accepting_new_bags
+                )
+            );
+        }
+    }
+}
+
+// Public methods
+impl<T: Trait> DataObjectStorage<T> for Module<T> {
+    fn can_upload_data_objects(params: &UploadParameters<T>) -> DispatchResult {
```

Comment:
Yes, it is possible. However, such an interface has a couple of downsides:
1. It would disclose several internal helper abstractions like BagChange, VoucherUpdate, BagManager, etc. It would complicate the pallet API with no obvious benefit.

2. There will be a way to call the methods with manually constructed data bypassing the actual checks.

3. It arguably looks like a contradiction with your other comment about 
`BagManager::<T>::ensure_bag_exists(&params.bag_id)?;`

```
.... 
This must mean that in many code paths you are going to ...<use this abstraction again> ...
...possibly with a fallible lookup, or a blind mutation, where the latter could risk updating an invalid (default constructed) instance.
```